### ISSUE TITLE
[angular/core] make DebugServices treeshakable (~10KB savings)

### ISSUE THREAD

**Issue by aitboudad** [angular/devkit#244]
### Bug Report or Feature Request (mark with an `x`)
```
- [ ] bug report -> please search issues before submitting
- [x] feature request
```

### Area
```
- [x] devkit
- [ ] schematics
```

### Desired functionality

Uglify is not able to remove unused debug services bocause of condition: [core.js#L13946](https://github.com/angular/core-builds/blob/5.0.0-rc.8/esm5/core.js#L13946) and replacing that condition with `createProdServices()` resulted in ~10KB savings

```patch
-- var /** @type {?} */ services = isDevMode() ? createDebugServices() : createProdServices();
++ var /** @type {?} */ services = createProdServices();
```

**Comment by IgorMinar** [comment]
This is very interesting but the implementation relies on a side effect of
the pure top level functions optimization. It could also be considered as a
breaking change in some scenarios. I'll look into this more deeply next
week.

**Comment by IgorMinar** [comment]
I've been thinking about this and it should be safe to build this optimization into build optimizer. @filipesilva what do you think?

**Comment by trotyl** [comment]
It would also save about 2.4KB in **Closure Compiler** when hard-coded `isDevMode` to `return false`.

Then comes an awkward situation, before we don't need to combine **Closure Compiler** with **Build Optimizer** since **Closure Compiler** could handle all cases **Build Optimizer** did. But that is not true any more.

**Comment by filipesilva** [comment]
The options to fix this on Build Optimizer are suboptimal I feel... because it compounds the CLI "environments" feature with the Angular `enableProdMode` feature. Let me explain this better.

In a default CLI app setup, doing `--prod` uses `--environment=prod` that has a (runtime) flag. That flag will, in default projects, activate `enableProdMode` in `main.ts`. So indeed the CLI could tell build optimizer that it was called with `--prod`, and build optimizer could strip out the debug services.

But what if the user uses other environment files? Or what if the user changes logic in `main.ts`? The `enableProdMode` call happens at runtime. Static analysis would be required at build time to determine whether the debug services should really be removed.

I think that https://github.com/angular/angular/issues/12200 would enable a fix for this issue as well, as it makes devMode a compile time concept instead of a runtime one.

**Comment by IgorMinar** [comment]
Yeah, you are right about this. I tagged that issue as v6 candidate.

**Comment by alexeagle** [comment]
This issue was moved to angular/angular-cli#12137
